repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/datamesh/DataProductPreProcessor.java

```diff
@@ -186,7 +218,127 @@ private static String createQualifiedName(String parentDomainQualifiedName) thro
         if (StringUtils.isEmpty(parentDomainQualifiedName)) {
             throw new AtlasBaseException(AtlasErrorCode.BAD_REQUEST, "Parent Domain Qualified Name cannot be empty or null");
         }
-        return parentDomainQualifiedName + "/product/" + getUUID();
+        return parentDomainQualifiedName + "/product/" + PreProcessorUtils.getUUID();
+
+    }
+
+    private void createDaapVisibilityPolicy(AtlasEntity entity,AtlasVertex vertex) throws AtlasBaseException {
+        String productGuid = vertex.getProperty("__guid", String.class);
+
+        AtlasEntity policy = new AtlasEntity();
+        policy.setTypeName(POLICY_ENTITY_TYPE);
+        policy.setAttribute(NAME, entity.getAttribute(NAME));
+        policy.setAttribute(QUALIFIED_NAME, productGuid + "/read-policy");
+        policy.setAttribute(ATTR_POLICY_ACTIONS, Arrays.asList("entity-read"));
+        policy.setAttribute(ATTR_POLICY_CATEGORY, MESH_POLICY_CATEGORY);
+        policy.setAttribute(ATTR_POLICY_TYPE, POLICY_TYPE_ALLOW);
+        policy.setAttribute(ATTR_POLICY_RESOURCES, Arrays.asList("entity:" + entity.getAttribute(QUALIFIED_NAME)));
+        policy.setAttribute(ATTR_POLICY_RESOURCES_CATEGORY, POLICY_RESOURCE_CATEGORY_PERSONA_ENTITY);
+        policy.setAttribute(ATTR_POLICY_SERVICE_NAME, "atlas");
+        policy.setAttribute(ATTR_POLICY_SUB_CATEGORY, DATA_PRODUCT); // create new constant attr
+
+        switch ((String) entity.getAttribute(DAAP_VISIBILITY_ATTR)) {
+            case PRIVATE:
+                setPolicyAttributes(policy, Arrays.asList(), Arrays.asList());
+                // do not create any auth policy in case of private daap visibility
+                break;
+            case PROTECTED:
+                setPolicyAttributes(policy,
+                        (List<String>) entity.getAttribute(DAAP_VISIBILITY_USERS_ATTR),
+                        (List<String>) entity.getAttribute(DAAP_VISIBILITY_GROUPS_ATTR)
+                );
+                createPolicy(policy);
+                break;
+            case PUBLIC:
+                setPolicyAttributes(policy, Arrays.asList(), Arrays.asList("public"));
+                createPolicy(policy);
+                break;
+        }
+    }
+
+    private void updateDaapVisibilityPolicy(AtlasEntity newEntity, AtlasEntity currentEntity,  String currentProductDaapVisibility, String newProductDaapVisibility) throws AtlasBaseException{
+        AtlasObjectId atlasObjectId = new AtlasObjectId();
+        atlasObjectId.setTypeName(POLICY_ENTITY_TYPE);
+        atlasObjectId.setUniqueAttributes(AtlasEntityUtils.mapOf(QUALIFIED_NAME,currentEntity.getGuid()+"/read-policy"));
+        AtlasVertex policyVertex = entityRetriever.getEntityVertex(atlasObjectId);
```

Comment:
Imagine case when moving visibility from Private to Protected/Public for the 1st time
In this case policyVertex will be null, if so we will need to generate a new Object of AtlasEntity similar to what we do in createProduct section